keygen/src/ssl.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2018 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/ecdsa.h>
+#include <openssl/x509.h>
+
+#include "ssl.h"
+
+/**
+ * gen_EC() - Generate an Elliptic Curve key (NID_X9_62_prime256v1)
+ *
+ * Return: A pointer to a new EC_KEY instance or NULL on failure
+ */
+EC_KEY * gen_EC()
+{
+    EC_KEY * ec_key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+
+    if (ec_key && !EC_KEY_generate_key(ec_key))
+    {
+        EC_KEY_free(ec_key);
+        ec_key = NULL;
+    }
+
+    return ec_key;
+}
+
+/**
+ * gen_cert() - Generate an X509 cert with a public key
+ * @pkey: The public key to certify as an EVP_PKEY*
+ * @cn: The IPv6 address (as a string) to assign to CN
+ *
+ * Return: A pointer to a new X509 certificate
+ */
+X509 *gen_cert(EVP_PKEY *pkey, char *cn)
+{
+    X509 *x509 = X509_new();
+
+    if(!x509)
+    {
+        goto out;
+    }
+
+    /* Set the serial number. */
```

Comment:
Sorry, I forgot to include this in my actual review.

We should actually generate a random serial number (both to help us track these certs, and also as general practice it looks bad to have all our certs have the same serial number).

So, we need something like
```
    unsigned char buffer[19];    // to hold a random number up to 2**152
    randombytes_buf(buffer, sizeof(buffer));
    BIGNUM *bn = BN_bin2bn(buffer, sizeof(buffer), NULL);
    if (!bn) {
        <handle error>
    }   

    ASN1_INTEGER *serial = BN_to_ASN1_INTEGER(bn, NULL);
    if (!serial) {
        <handle error>
    } 

    if (!X509_set_serialNumber(x509, serial)) {
        <handle error>
    }
```

This requires that we `#include <sodium.h>`. This now uses the `libsodium` library for generating the random serial number; that should be fine, since `enftun` already has to link against libsodium for `xtt`.

Reply:
Done